io-sim/src/Data/List/Trace.hs

```diff
@@ -65,6 +69,32 @@ ppTrace :: (a -> String) -> (b -> String) -> Trace a b -> String
 ppTrace sa  sb (Cons b bs) = sb b ++ "\n" ++ ppTrace sa sb bs
 ppTrace sa _sb (Nil a)     = sa a
 
+-- | Take the first n elements of a Trace, converting each to ().
+take :: Int -> Trace a b -> Trace (Maybe a) b
+take n _ | n <= 0 = Nil Nothing
+take _ (Nil a)    = Nil (Just a)
+take n (Cons b o) = Cons b (take (n - 1) o)
+
+-- | Take elements from the Trace while the predicate holds, converting each to ().
+takeWhile :: (b -> Bool) -> Trace a b -> Trace (Maybe a) b
+takeWhile _ (Nil a)          = Nil (Just a)
+takeWhile p (Cons b o)
+  | p b                       = Cons b (takeWhile p o)
+  | otherwise                 = Nil Nothing
```

Comment:
```suggestion
takeWhile _ (Nil a)    = Nil (Just a)
takeWhile p (Cons b o)
  | p b                = Cons b (takeWhile p o)
  | otherwise          = Nil Nothing
```